docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,286 @@
+  twin for this one on the LNS side. For this the Function should perform the Mic computation which
+  means that we would need to send the payload to the cloud. This is a deal breaker for us.
+
+#### Further processing
+
+This section uses this topology:
+
+```mermaid
+flowchart LR;
+    Device-->LBS1-->LNS1-->Function;
+    Device-->LBS2-->LNS2-->Function;
+    LNS1-->IoTHub;
+    LNS2-->IoTHub;
+```
+
+where Device sends data message A and then B.
+
+Here is a rundown of what should happen assuming both LNSs have a fresh twin (via background refresh
+or via fetching it using DeviceGetter.GetDevice). Changes are marked in **bold**:
+
+1. Device sends first data message A.
+1. We assume that LNS1 gets the message first. **LNS1 checks against
+   its in-memory dictionary DevEui -> flag** and since it has not seen this DevEui before contacts the Function.
+1. The Function hasn't seen this DevEui either and therefore does not have an assigned LNS for it
+  yet. LNS1 wins the race and gets immediately a response and processes the message upstream.
+1. LNS2 eventually receives message A, **checks its local dictionary** and also contacts the
+   FunctionBundler immediately since it does not have prior info about this DevEui.
+1. The Function responds to LNS2 that it lost the race to process this message.
+1. Since deduplication strategy is Drop, LNS2 drops the message immediately, therefore no
+   connection to Iot Hub is opened and only LNS1 has the connection to Iot Hub. **LNS2 notes in
+   memory that it was the losing gateway for this DevEui**.
+1. When message B gets send (with a higher frame counter*), assuming that this time LNS2 gets it
+   first it **checks again its local dictionary it's not the preferred LNS for this device and
+   therefore delays itself X ms before contacting the FunctionBundler**.
+   - Here we do *not* want to simply drop the message as LNS1 might not be available anymore (due to
+     a crash, device not in range etc).
+   - Since it was the losing LNS previously it should **make sure that it operates on a fresh twin**
+     (especially regarding the frame counter) see also [background refresh section](#handling-of-cache-refresh)
+1. This delay gives LNS1 a time advantage to reach the FunctionBundler first and win the race again, failing
+    back to the previous case of message A. The active connection stays with LNS1.
+1. If this delay is not sufficient for LNS1 to win the race, LNS2 will contact the FunctionBundler which
+   now awards LNS2 as the "winning" LNS. LNS2 can now process the message upstream. It also **removes the "losing flag" from its in-memory store**.
+1. **The Function also proactively informs LNS1** that it's not anymore the winning LNS for this device.
+   - The reason why we do this proactively is that otherwise its Edge Hub will try to reconnect to IoT Hub even if there is no more messages picked up from LNS1 (out of range roaming client).
```

Comment:
```suggestion
   - The reason why we do this is to ensure, LNS1 knows connection ownership was transferred to another network server and it can drop the connection. This is for the case, where the upstream message does not reach LNS1.
```